### PR TITLE
[NFC] Standardize dump() methods in frontend

### PR DESCRIPTION
**Pull request by beccadax**
By convention, most structs and classes in the Swift compiler include a `dump()` method which prints debugging information. This method is meant to be called only from the debugger, but this means they’re often unused and may be eliminated from optimized binaries. On the other hand, some parts of the compiler call `dump()` methods directly despite them being intended as a pure debugging aid. clang supports attributes which can be used to avoid these problems, but they’re used very inconsistently across the compiler.

This commit adds `SWIFT_DEBUG_DUMP` and `SWIFT_DEBUG_DUMPER(<name>(<params>))` macros to declare `dump()` methods with the appropriate set of attributes and adopts this macro throughout the frontend. It does not pervasively adopt this macro in SILGen, SILOptimizer, or IRGen; these components use `dump()` methods in a different way where they’re frequently called from debugging code. Nor does it adopt it in runtime components like swiftRuntime and swiftReflection, because I’m a bit worried about size.

Despite the large number of files and lines affected, this change is NFC.

### CODE REVIEW

**Comment by brentdax** [comment]
@swift-ci please smoke test compiler performance

**Comment by brentdax** [comment]
@swift-ci please smoke test

**Comment by brentdax** [comment]
Some specific things I'd like opinions on:

* Is this a good idea at all?
* `DECLARE_DUMP()` and `DECLARE_DUMP_NAMED()`: Good names, bad names, or ugly names?
* Should the macros make the methods `const`? This seems to force some `const_cast`ing in certain places.

File: include/swift/AST/AnyFunctionRef.h [hunk 1]
```diff
@@ -177,8 +177,7 @@ class AnyFunctionRef {
 #pragma warning(push)
 #pragma warning(disable: 4996)
 #endif
-  LLVM_ATTRIBUTE_DEPRECATED(void dump() const LLVM_ATTRIBUTE_USED,
-                            "only for use within the debugger") {
+  DECLARE_DUMP() {
```
Comment (jrose-apple) [inline include/swift/AST/AnyFunctionRef.h]: This one spooked me. Now it's not just a declaration!

File: include/swift/Basic/LLVM.h [hunk 2]
```diff
@@ -89,4 +89,16 @@ namespace swift {
...
+/// dumping. Methods declared with this macro should never be used except in the
+/// debugger; they are deprecated and marked used.
+#define DECLARE_DUMP_NAMED(name, ...) \
+LLVM_ATTRIBUTE_DEPRECATED(void name(__VA_ARGS__) const LLVM_ATTRIBUTE_USED, \
+                          "only for use in the debugger")
+
+/// Declares a const method named `dump` with the given arguments.
+/// Methods declared with this macro should never be used except in the
+/// debugger; they are deprecated and marked used.
+#define DECLARE_DUMP(...) DECLARE_DUMP_NAMED(dump, __VA_ARGS__)
```
Comment (jrose-apple) [inline include/swift/Basic/LLVM.h]: These don't belong in swift/Basic/LLVM.h; this header is only for bringing LLVM things into Swift's namespace. Maybe just make a new one?

**Review by jrose-apple — COMMENTED** [review]
I think this is net good (and it helped you find a bunch of dumps using inconsistent streams, which is good). I'm not 100% sold on the name; at the very least it needs a `SWIFT` prefix, but it's also not always a declaration. Maybe just `SWIFT_DUMP_METHOD`?

File: include/swift/Basic/LLVM.h (hunk 2 above)
Comment (brentdax) [reply include/swift/Basic/LLVM.h]: My thought was that this is wrapping `LLVM_ATTRIBUTE_DEPRECATED` and `LLVM_ATTRIBUTE_USED`, but that's kind of an implementation detail—I'll look for a better home for it.

File: include/swift/Basic/LLVM.h [hunk 3]
```diff
@@ -89,4 +89,16 @@ namespace swift {
   using llvm::raw_ostream;
 } // end namespace swift
 
+/// Declares a const method with the given name and arguments intended for
+/// dumping. Methods declared with this macro should never be used except in the
+/// debugger; they are deprecated and marked used.
+#define DECLARE_DUMP_NAMED(name, ...) \
```
Comment (brentdax) [inline include/swift/Basic/LLVM.h]: This apparently doesn't work with empty parameter lists in the host compilers on Linux bots, and I'm not sure MSVC will take it either. I'll probably change it to something with a use site like (ignoring renaming):

```
DECLARE_DUMP_NAMED(dumpContext());
```

Comment (jrose-apple) [reply include/swift/Basic/LLVM.h]: Ooh, `void` will work.

**Comment by brentdax** [comment]
Changes in this push:

* The macros are now in swift/Basic/Debug.h.
* They are now called `SWIFT_DEBUG_DUMP(<params>)` and `SWIFT_DEBUG_DUMPER(<name>(<params>))`. For additional flexibility, a new `SWIFT_DEBUG_HELPER(signature)` macro just adds the deprecation and used attributes without imposing anything else on the declaration.
* I've mentioned these in docs/DebuggingTheCompiler.rst. I'll probably also post on Swift Forums once this is in.
* I've converted a few uses in the SIL phase that *do* match the Sema convention to use the new macros.

**Comment by brentdax** [comment]
@swift-ci please smoke test

File: include/swift/Basic/Debug.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+/// Methods declared with this macro should never be used except in the
+/// debugger.
+#define SWIFT_DEBUG_DUMPER(nameAndParams) \
+  SWIFT_DEBUG_HELPER(void nameAndParams const)
+
+/// Declares a const void instance method named `dump` with the given
+/// parameters. Methods declared with this macro should never be used except in
+/// the debugger.
+#define SWIFT_DEBUG_DUMP(...) \
+  SWIFT_DEBUG_DUMPER(dump(__VA_ARGS__))
```
Comment (jrose-apple) [inline include/swift/Basic/Debug.h]: I thought this was the pattern that wasn't allowed to be empty. :-/

Comment (jrose-apple) [reply include/swift/Basic/Debug.h]: C++11 [cpp.replace]p4: "[…] there shall be more arguments in the invocation than there are parameters in the macro definition (excluding the `...`)"

Comment (jrose-apple) [reply include/swift/Basic/Debug.h]: You could get around this by having `SWIFT_DEBUG_DUMP` take no arguments and making the few `dump` methods that take arguments go through `SWIFT_DEBUG_DUMPER` instead?

**Comment by brentdax** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckConstraints.cpp [hunk 5]
```diff
@@ -3617,12 +3617,12 @@ void Solution::dump(raw_ostream &out) const {
   }
 }
 
-void ConstraintSystem::dump() {
-  print(llvm::errs());
+void ConstraintSystem::dump() const {
+  const_cast<ConstraintSystem*>(this)->print(llvm::errs());
```
Comment (CodaFi) [inline lib/Sema/TypeCheckConstraints.cpp]: I think you can drop these by marking `ConstraintSystem::print` const.

Comment (brentdax) [reply lib/Sema/TypeCheckConstraints.cpp]: Took a bit of doing, but I did it. I've also squashed the initial commits.

Comment (brentdax) [reply lib/Sema/TypeCheckConstraints.cpp]: Took a bit of doing, but I did it. I've also squashed the initial commits.

**Comment by brentdax** [comment]
Latest force-pushes:

* Removed `const` from some dumpers so they didn't need `const_cast`s.
* Added `const` to callees of other dumpers so they also didn't need `const_cast`s.
* Cleaned up a couple of small messes, like `dump(llvm::errs())` turning into `this->dump(llvm::errs())` in a couple spots.

**Comment by brentdax** [comment]
@swift-ci please smoke test

**Comment by brentdax** [comment]
Since I've already failed on one OS due to compiler macro limitations:

@swift-ci Please test Windows platform

**Comment by davezarzycki** [comment]
How does this compile? I'm trying to test swift's master branch (17169fc1fe4157a4ac61a3ba36cde209ca8f54bd) with swift-master from llvm-project (4f5a37c2d53f37fb6b7b4eabb993eb5844ed5b71). What am I missing?
```
FAILED: tools/swift/lib/IRGen/CMakeFiles/swiftIRGen.dir/GenProto.cpp.o 
/p/llvm/bin/clang++  -DGTEST_HAS_RTTI=0 -D_DEBUG -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Itools/swift/lib/IRGen -I/home/dave/s/u/swift/lib/IRGen -Itools/swift/include -I/home/dave/s/u/swift/include -I/usr/include/libxml2 -Iinclude -I/home/dave/s/u/llvm/include -Itools/clang/include -I/home/dave/s/u/llvm/../clang/include -I/home/dave/s/u/cmark/src -Itools/cmark/src -Wno-c11-extensions -Wno-c99-extensions -Wno-c++17-extensions -Wno-dollar-in-identifier-extension -Wno-empty-translation-unit -Wno-format-pedantic -Wno-nullability-extension -Wno-four-char-constants -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -std=c++14 -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wstring-conversion -fdiagnostics-color -ffunction-sections -fdata-sections -Wno-nested-anon-types -Werror=switch -Wdocumentation -Wimplicit-fallthrough -Wunreachable-code -Woverloaded-virtual -DOBJC_OLD_DISPATCH_PROTOTYPES=0 -O2    -UNDEBUG  -march=native -fno-omit-frame-pointer -fno-vectorize -fno-slp-vectorize -mno-omit-leaf-frame-pointer -fno-exceptions -fno-rtti -target x86_64-unknown-linux-gnu -O2 -g0 -UNDEBUG  -Werror=gnu -MD -MT tools/swift/lib/IRGen/CMakeFiles/swiftIRGen.dir/GenProto.cpp.o -MF tools/swift/lib/IRGen/CMakeFiles/swiftIRGen.dir/GenProto.cpp.o.d -o tools/swift/lib/IRGen/CMakeFiles/swiftIRGen.dir/GenProto.cpp.o -c /home/dave/s/u/swift/lib/IRGen/GenProto.cpp
In file included from /home/dave/s/u/swift/lib/IRGen/GenProto.cpp:60:
In file included from /home/dave/s/u/swift/lib/IRGen/Fulfillment.h:24:
In file included from /home/dave/s/u/swift/lib/IRGen/MetadataPath.h:21:
In file included from /home/dave/s/u/swift/include/swift/Basic/EncodedSequence.h:26:
/home/dave/s/u/swift/include/swift/Basic/PrefixMap.h:652:21: error: invalid operands to binary expression ('llvm::raw_ostream' and 'swift::irgen::MetadataResponse')
        out << " (" << node->get() << ')';
        ~~~~~~~~~~~ ^  ~~~~~~~~~~~
/home/dave/s/u/swift/include/swift/Basic/PrefixMap.h:645:22: note: in instantiation of member function 'swift::PrefixMap<unsigned char, swift::irgen::MetadataResponse, 7>::print' requested here
  SWIFT_DEBUG_DUMP { print(llvm::errs()); }
                     ^
/home/dave/s/u/swift/include/swift/Basic/EncodedSequence.h:351:13: note: in instantiation of member function 'swift::PrefixMap<unsigned char, swift::irgen::MetadataResponse, 7>::dump' requested here
    MapBase TheMap;
            ^
```

Followed by tons of various `<<` functions being not viable. For example:
```
/home/dave/s/u/llvm/include/llvm/ADT/Optional.h:415:14: note: candidate function not viable: no known conversion from 'swift::irgen::MetadataResponse' to 'llvm::NoneType' for 2nd argument
raw_ostream &operator<<(raw_ostream &OS, NoneType);
             ^
```

**Comment by davezarzycki** [comment]
@brentdax – I think I figured out the bug.

@bob-wilson  – Does Apple's clang and Ubuntu's clang predate https://reviews.llvm.org/D58216 ?

@rjmccall – It seems that `irgen::MetadataResponse` lacks `operator<<`. Is that correct?

**Comment by davezarzycki** [comment]
The fix: #28034

**Comment by rjmccall** [comment]
> @rjmccall – It seems that `irgen::MetadataResponse` lacks `operator<<`. Is that correct?

If it's not there, it was probably just never useful enough to include.

**Pull request merged**